lint-rules.md

```diff
@@ -0,0 +1,1280 @@
+
+We are aiming to automate as much as we can to make adherence to these lint rules as easy as possible. An "auto-formatter" will be packaged as part of our [`sprocket` tool](https://github.com/stjude-rust-labs/sprocket). Auto-formatting has not yet been implemented as this project is still early in development, but while reading this document, please assume that anything easily automated (and perhaps complex to do manually) will eventually be handled by the formatter built into `sprocket`.
+
+Some of these lint warnings are arbitrary. There are times where several different methods for styling a document could be considered equally appealing. The important part is picking one such style and adhering to it consistenly across a document or codebase. In such cases, this document will only present one option. In the future, that singular option may just be the default, with several more-or-less equivalent styles presented as alternative options available through configuration. If you feel we are picking the wrong defaults, please let us know what you think would be better!
+
+At the time of writing (January 2024), we consider all rules herein to be up for debate. Please open an issue on this repository if you would like to suggest changes to this document.
+
+## Rules
+
+### `version_declaration_placement`
+
+The WDL version declaration is required for any v1 WDL document to be parsed. A missing version declaration will result in a validation error. However an incorrect placement of the decleration will be considered a lint warning. The version decleration should be the very first line in the document, unless there are header comments. In which case, there should be a blank line between the header and the version declaration. There should _always_ be a blank line following the version declaration.
+
+**Group**: `spacing`
+
+#### Example
+
+Good:
+
+```wdl
+version 1.1
+
+...
+```
+
+Also good:
+
+```wdl
+## [Homepage](https://example.com/)
+
+version 1.0
+
+...
+```
+
+### `import_placement` && `import_sort`
+
+All import statements should follow the WDL version declaration (with one empty line between the version and the first import statement).
+
+Import statements should be sorted by the lexicographical ordering (GNU `sort` with `LC_COLLATE=C`) of each line. No extra white space is allowed between symbols or lines.
+
+`import_placement` **group**: `spacing`
+
+`import_sort` **group**: `sorting`
+
+#### Example
+
+Good:
+
+```wdl
+version 1.1
+
+import "../../tools/fastqc.wdl" as fastqc_tasks
+import "../../tools/fq.wdl"
+import "./markdups-post.wdl" as markdups_post_wf
+import "https://raw.githubusercontent.com/stjude/seaseq/3.0/workflows/tasks/seaseq_util.wdl" as seaseq_util
+```
+
+### `blanks_between_elements`
```

Comment:
- (Strong opinion) When `call`ing a task in a workflow, you should format it so that the `input:` line is indented on the next line alone, and then indent again for the input assignments, each on one line. So

```
   call greet { input:
        name = name,
        greeting = greeting,
    }
```
should be
```
call greet { 
    input:
        name = name,
        greeting = greeting,
    }
```
- This case should maybe be a separate lint about styling task calls. (See also thoughts on `input_spacing` below)
- Also (maybe) consider removing whitespace between input assignments like in Python function calling with keyword assignments, e.g. `name=name,` etc. (though this conflicts with `expression_spacing` later... maybe a separate lint for spacing within call blocks? But I don't care as much about this detail honestly)

Reply:
For communication purposes, let's refer to the style currently proposed by `lint-rules.md` as "inline-input" and the style you proposed as "indented-input". (Accepting alternative names)

I've seen indented-input in several repos. Do you think this is the style the community is coalescing around?

I strongly prefer inline-input. I think good universal rules are that opening brackets should be on the same indent level as their closing bracket, and that code should never "skip" an indent level by incrementing or decrementing by greater than 1 level at a time. Both of these rules are broken by indented-input.

Why do you prefer indented-input? 

-------------------

> * Also (maybe) consider removing whitespace between input assignments like in Python function calling with keyword assignments, e.g. `name=name`, etc. (though this conflicts with `expression_spacing` later... 

First, I find the whitespace padding on assignments easier to read. I wouldn't throw a hissy fit over losing that whitespace though if my opinion is the minority.

Second, the lints are written such that the `input_spacing` lint and the `expression_spacing` lints can "contradict" (though that may prove confusing). By that I mean we _could_ have a lint active for "no spaces around input assignments (`=`)" **and also** have a lint active for "whitespace padding around expression assignments (`=`)". They are different contexts. However my gut is that they should treat whitespace the same way around _all_ equal signs for simplicity.

Reply:
I agree with your second point. I've just gotten into the habit of no surrounding whitespace for `call`s since Python has the same weird dichotomy e.g. `x = 4` on its own but `my_func(x=4, y=7)` for function calls. WDL can be different and more consistent with whitespace everywhere.

---
I don't have a sense of the "larger community" and just have a biased view of the WDLs I've seen around Broad. But I never saw inline-input style until I read this lint guide. Here's an argument in favor of indented-input. What is the purpose of the syntax `input:` at all instead of e.g.
```
call MyTask {
    file=file,
    var=var,
}
```
Currently this syntax is conceptually OK, but for whatever reason the `input:` is required. This suggests there could be a future where other blocks of variables are fed to a task call with different headers, e.g.
```
File my_task_log_file

call MyTask {
    input:
        file=file,
        var=var,
    debug:
        log_file=my_task_log_file
}
```
In this case, using inline-input has no natural way to extend to the new `debug:` situation, and would look very ugly. Perhaps it's worth asking some WDL developers what the intent behind including `input:` in `call` blocks at all is to help determine how to style it.

Reply:
https://github.com/openwdl/wdl/pull/635/files

In the 1.2 WDL release they are going to remove the requirement for the `input:` token. When that time comes, I am _definitely_ voting for deleting all uses of the token and updating the lint rules to match. With that said, I think we should adopt the style which will require the least amount of work when 1.2 is released. That would be the "inline-input" I originally proposed. A simple `sed -i` would be all that's needed to mass migrate to 1.2 style. It would be much more complicated to deal with the indentation changes required by the indented-input.

Reply:
I definitely completely agree that when it becomes obsolete it should be linted away. I asked in Slack to confirm that this means there definitely will never be additional headers in the `call` blocks, since I could imagine the developers choosing that `input` is optional when it's the only header, but also still support multiple headers in the future where the `input` becomes mandatory to specify it. Given the possibility for this (at the current stage), I still think the indented-input is more flexible to handle these (and am obviously biased since I use that). I wouldn't worry about refactoring it away in the future as a reason to choose either: whatever autofixing lint command that's written shouldn't struggle over a find/replace within `call` blocks vs also fiddling some whitespace. If a "vote" does happen by scraping repos I'd be curious to see which of these styles (or even a third?!) end up being most common in the wild.

Reply:
> Given the possibility for this (at the current stage), I still think the indented-input is more flexible to handle these (and am obviously biased since I use that).

I don't think it makes sense to choose our style based on a direction the SPEC may or may not take. Especially because that _potential_ change would necessitate new style lints if it ever came to fruition. i.e. if in the future they add non-input blocks to the call, we could just as easily change from inline-input to indented-input to some new 3rd option. Which brings me around to agreeing with your point:
> whatever autofixing lint command that's written shouldn't struggle over a find/replace within call blocks vs also fiddling some whitespace.

So I'll agree the point I made in the above comment wasn't very convincing lol. But the same logic applies to your comment as well.

>  If a "vote" does happen by scraping repos I'd be curious to see which of these styles (or even a third?!) end up being most common in the wild.

I proposed that repo scraping idea, but now that I've thought about it some more, it seems to fall flat. As you (@rickymagner) pointed out, the "sampling" will over-represent certain styles such as the Broad. I don't think what I proposed would lead to anything representative.

I think much less over-engineered way to go about this will be a simple Google poll or something like that which gets shared around. Or maybe even the lowest tech option would be a poll _within_ the openwdl Slack. Regardless, I think it's a bridge we don't need to cross now.

Reply:
I agree something like a carefully thought out poll on a few lint decisions that gets spread around different groups somehow fairly sounds like a good idea. There's always risk of oversampling either way (I'd definitely advertise around Broad which could lead to oversampling/bias). There's also a possibility for divisive lints to come in non-compatible either-or groups, like an inline vs indented input in calls, etc.